### PR TITLE
Additions on h-levels

### PR DESCRIPTION
**Pull request by IanRay11**

### CODE REVIEW

**Comment by tomdjong** [comment]
Some context: This is the first of three PRs on truncations and connectedness as I've asked @IanRay11 to split up his work.

**Review by tomdjong — CHANGES_REQUESTED** [review]
Thanks @IanRay11! :smile: 

Some additional comments that I couldn't make in place because these lines were not touched by the commit:

- Can you use `＝-retract` (from `UF.Retracts` and which I've just added) in `level-closed-under-retract` and fix the indentations to a single space in all the `hlevel-closed-under-`lemmas.

- In `hlevel-closed-under-retract`, `hlevel-closed-under-equiv` and `hlevel-closed-under-embedding`, the types should probably be implicit arguments.

- In `hlevel-closed-under-embedding`, the argument `n-above-one` isn't used, so you can do:
```
hlevel-closed-under-embedding (succ n) _ X Y (e , is-emb) Y-h-level x x' =
 hlevel-closed-under-equiv n (x ＝ x') (e x ＝ e x')
                           (ap e , embedding-gives-embedding' e is-emb x x')
                           (Y-h-level (e x) (e x'))
```

- In the equivalence reasoning of `ℍ-is-of-next-hlevel`, please put long long equivalences in a where clause, so that it looks like this
```
 X ≃⟨ I ⟩
 Y ≃⟨ II ⟩
 Z ■

 where
  I = some-long-equivalence
  II = some-other-long-equivalence
```

- Also in `ℍ-is-of-next-hlevel`, you can also just write `fe'` instead of `fe 𝓤 𝓤` and use `Π₂-is-prop` instead of `Π-is-prop` twice.

The comment about `fe'` applies elsewhere too.

File: source/UF/H-Levels.lagda [hunk 1]
```diff
@@ -33,9 +34,12 @@ open import UF.Subsingletons-FunExt
 open import UF.Subsingletons-Properties
 open import UF.Univalence
 open import UF.UA-FunExt
+open import Naturals.Addition renaming (_+_ to _+'_)
 open import Naturals.Order
 
-module UF.H-Levels (fe : FunExt) (fe' : Fun-Ext) where
+module UF.H-Levels (fe : FunExt)
+                   (fe' : Fun-Ext)
```
[the remark this answers is not in the document]
Comment (IanRay11) [reply source/UF/H-Levels.lagda:41]: I removed the FunExt assumption and made fe : Fun-Ext. Then \ _ _ -> fe : FunExt. Is this sufficient? I couldn't find the proof of interderivability in the FunExt file.

File: source/UF/H-Levels.lagda [hunk 2]
```diff
@@ -48,27 +52,38 @@ hlevel-relation-is-prop {𝓤} (succ n) X =
...
+map f is-of-hlevel n = (y : codomain f) → (fiber f y) is-of-hlevel n
+
 \end{code}
 
 H-Levels are cumulative.
 
 \begin{code}
 
+contr-implies-id-contr : {X : 𝓤 ̇} → is-contr X → (x x' : X) → is-contr (x ＝ x')
+contr-implies-id-contr (c , C) x x' = (((C x)⁻¹ ∙ C x') , D)
```
[the remark this answers is not in the document]
Comment (IanRay11) [reply source/UF/H-Levels.lagda:65]: Instead I am going to move the is-prop' introduction up.

File: source/UF/H-Levels.lagda (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomdjong) [reply source/UF/H-Levels.lagda:41]: It's standard TypeTopology practice to do something like
```
private
 fe' : Fun-Ext
 fe' {𝓤} {𝓥} = fe 𝓤 𝓥
 ```

[the remark this answers is not in the document]
Comment (tomdjong) [reply source/UF/H-Levels.lagda:41]: It really doesn't require a proof as the only difference is having the universes implicit or not.

**Pull request closed without merging**